content/learning-paths/mobile-graphics-and-gaming/android_halide/android.md

```diff
@@ -0,0 +1,454 @@
+        }
+        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
+        return bitmap
+    }
+
+    external fun blurThresholdImage(inputBytes: ByteArray, width: Int, height: Int): ByteArray
+
+    companion object {
+        // Used to load the 'armhalideandroiddemo' library on application startup.
+        init {
+            System.loadLibrary("armhalideandroiddemo")
+        }
+    }
+}
+```
+
+This Kotlin Android application demonstrates integrating a Halide-generated image-processing pipeline within an Android app. The main activity (MainActivity) manages loading and processing an image stored in the application’s asset folder.
+
+When the app launches, the Process Image button is disabled. When a user taps Load Image, the app retrieves img.png from its assets directory and displays it within the ImageView, simultaneously enabling the Process Image button for further interaction.
+
+Upon pressing the Process Image button, the following sequence occurs:
+1. Background Processing. A Kotlin coroutine initiates processing on a background thread, ensuring the application’s UI remains responsive.
+2. Conversion to Grayscale. The loaded bitmap image is converted into a grayscale byte array using a simple RGB-average method, preparing it for processing by the native (JNI) layer.
+3. Native Function Invocation. This grayscale byte array, along with image dimensions, is passed to a native function (blurThresholdImage) defined via JNI. This native function is implemented using the Halide pipeline, performing operations such as blurring and thresholding directly on the image data.
+4. Post-processing. After the native function completes, the resulting processed grayscale byte array is converted back into a Bitmap image.
+5. UI Update. The coroutine then updates the displayed image (on the main UI thread) with this newly processed bitmap, providing the user immediate visual feedback.
+
+The code defines three utility methods:
+1. loadImageFromAssets, which retrieves an image from the assets folder and decodes it into a Bitmap.
+2. extractGrayScaleBytes - converts a Bitmap into a grayscale byte array suitable for native processing.
+3. createBitmapFromGrayBytes - converts a grayscale byte array back into a Bitmap for display purposes.
+
+Note that performing the grayscale conversion in Halide allows us to exploit operator fusion, further improving performance by avoiding intermediate memory accesses. This could be done as follows:
+```cpp
+// Halide variables
+Halide::Var x("x"), y("y"), c("c");
+
+// Original RGB input buffer (interleaved RGB)
+Halide::Buffer<uint8_t> inputBuffer(inputRgbData, width, height, 3);
+
+// Convert RGB to grayscale directly in Halide pipeline
+Halide::Func grayscale("grayscale");
+grayscale(x, y) = Halide::cast<uint8_t>(
+    0.299f * inputBuffer(x, y, 0) +
+    0.587f * inputBuffer(x, y, 1) +
+    0.114f * inputBuffer(x, y, 2)
+);
+
+// Continue pipeline: Gaussian blur (example)
```

Comment:
Please write consistent code for the same thing throughout this LP as long as it makes sense.
I mean to say here specifically is that the halide code below is slightly different from what we presented in the previous chapter.